### ISSUE TITLE
[KNX] bad connection handling and resulting memory leak

### ISSUE THREAD

**Issue by muenning** [openhab/openhab1-addons#5440]
OpenHAB suffers a major memory leak caused by losing threads probably started by the KNX binding. 

## Expected Behavior
I expect OpenHAB to run reliable without the need to restart it several times a day.

## Current Behavior
After some running (12-24h with OpenHAB 1.8.3 and 4-12h with 2.x) the java thread count reaches its limit of about 8000, messages like `java.lang.OutOfMemoryError: unable to create new native thread` are filling the log and OpenHAB to basically stop working. After a restart everything is fine until the thread spawning starts again.

After some searching I think I have pinpointed the KNX binding to be the source of the problem. For unclear reasons the connection to the KNX router (MDT IP Router SCN-IP100.02 connected in TUNNEL mode as ROUTER mode is not working) is closed with a message like this logged:

```
2018-01-17 18:13:11.180 [WARN ] [.binding.knx.internal.bus.KNXBinding] - Value 'OFF' could not be sent to the KNX bus using datapoint 'command DP 10/3/56 WRNW, DPT main 0 id 1.001, low priority' - retrying one time: link closed, waiting for service ack
2018-01-17 18:13:11.180 [WARN ] [.binding.knx.internal.bus.KNXBinding] - Value 'OFF' could not be sent to the KNX bus using datapoint 'command DP 10/3/56 WRNW, DPT main 0 id 1.001, low priority' - retrying one time: link closed, connection closed
2018-01-17 18:13:11.179 [ERROR] [tuwien.auto.calimero                ] - calimero.link.192.168.0.16:3671: send error, closing link
```

After the first disconnect a continuous stream of connect attempts and failures are logged:

```
2018-01-17 18:16:10.066 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
2018-01-17 18:16:10.108 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:16:10.109 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:16:10.113 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:10.114 [INFO ] [nx.internal.connection.KNXConnection] - Connected to KNX
2018-01-17 18:16:57.922 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
2018-01-17 18:16:58.110 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:16:58.113 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:16:58.114 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-17 18:16:59.526 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:59.529 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:59.530 [INFO ] [nx.internal.connection.KNXConnection] - Connected to KNX
2018-01-17 18:17:24.837 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:17:35.908 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:17:28.115 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
```

This leads to one thread left over on every connection attempt. As the threads are not named I could verify this only by counting the log entries and comparing to the increase of java threads in the system. These connection attempts also use up all available channels on the KNX Router - ETS complains that there are no available connections left so I have to stop OpenHAB if I want to get a connection. Before this excessive re-connection attempts I can use ETS in parallel to running OpenHAB as expected.

I could not see any change in the behavior regardless of the `numberOfThreads` setting in knx.cfg. Basically no setting seems to have any effect on this.

By the way the disconnect message(s) are kinda bogus as KNX communication continues to work as expected even if OpenHAB pretends to be disconnected and tries to reconnect.

## Possible Solution
Check thread handling in KNX binding in regard of what happens when a connection is lost and retried. Terminating the left-over threads before reusing their handles should do the trick. Or some calimero clean-up is missed or the connection status is not queried correctly.
On the other hand re-sending the values to the bus seems to trigger the connection issue or at least makes it more probable. These resends happen regardless the `ignorelocalevents=true` setting.

## Steps to Reproduce (for Bugs)
Just let it run and watch the results. As this is probably related to the KNX hardware used (MDT IP Router SCN-IP100.02) and the amount of traffic (more than 100 KNX devices) it could be difficult to reproduce it on another or smaller system. On a smaller system it may just need to run longer to happen.

## Your Environment
openhab-1.8.3 with knx 1.9.0 binding
(all OpenHAB versions in between have the same problem)
openhab-2.2.0 with knx 1.11.0 binding
IP Router MDT SCN-IP100.02 connected in TUNNEL mode

my knx.cnf:
```
ip=192.168.0.16
ignorelocalevents=true
type=TUNNEL
localIp=192.168.0.1
autoReconnectPeriod=30
numberOfThreads=1
```

**Comment by ThKern** [comment]
I can confirm exactly the same behavior for me, also openhab 2.2 latest and knx plugin. In fact I just wanted to open the same issue...
Regards
Thomas

**Comment by muenning** [comment]
I have this problem several months now (since before 2.1 was released, from the very beginning I started to use KNX binding) and I was curious if anyone else had it - there were no bug reports about this and the software updates/new versions did not change anything. Even going from OpenHAB 1.x to 2.x made things worse. @ThKern, is your KNX installation also rather large or why do we have the problem and others seems to not?

My current experience is that using OpenHAB 1.x is most stable as it takes at least twice the time until it becomes unusable. Still not a solution, though.

**Comment by ThKern** [comment]
~2 years back I used OpenHAB 1.x until they released openHAB 2. I ran this on a synology NAS. After switching to openHAB 2 I changed so much that there is no way back to 1.x. And then these memory leaks started. As I also did not find any other reports I assumed that they are owed to the pretty outdated Java installation (I have a DS413, bad choice, as the processor is too exotic so that Oracle stopped updates long ago). I gave up openHAB completely as in this case the NAS became unusable because of the frequent Java dumps, and openHAB did not do his work anymore, like rules etc.
Now I took a new start with a RPi 3 and openHAB 2.2. This time I only observed that the connections are all eaten up, but no out-of-mem so far, and no Java dumps -> the openHAB server stays usable at least.
so this is for me the least bad solution. I am now thinking about a cron job on my RPi which will run systemctl restart openhab2.service each night at 3. That should help, the short downtime should not have negative impact, as all are asleep, people and house...Have you tried this?
I have not sooo many datapoints. I suspected in the beginning that only if you change some .item or .thing file the problem exists, but that does not seem to be true, it also happens if you leave the openHAB settings completely in peace.

**Comment by kaikreuzer** [comment]
FTR: I am using the KNX1 binding since 2010 from openHAB 0.1.0 to 2.3.0-SNAPSHOT.
I have never encountered any memory issues from this binding, but in fact, I also have no connection issues (using a single connection Siemens N148 IP interface).

The problems is therefore indeed the reconnections that you should best solve for your setup.

When testing the KNX2 binding recently, I came across a bug that did several reconnects a second and I immediately hit the same problem as you which you can see in https://github.com/openhab/openhab2-addons/pull/2323#issuecomment-356052455.

This is clearly a bug in the Calimero library (@sjka Did you already enter a bug report for that and have a reference?). This library was [last changed in Oct 2016 in the KNX binding](https://github.com/openhab/openhab1-addons/tree/master/bundles/binding/org.openhab.binding.knx/lib), so I would assume that the problem exists at least since then.
@lewie As you are a heavy user, did you encounter this issue already? Any advice on how to work around it?

**Comment by ThKern** [comment]
Oct 2016: Definitely the problem began earlier (for me), see my [cry-for-help](https://community.openhab.org/t/most-likely-memory-leak-how-to-identify-the-problem/24619) from Mar 1st 2017...But at that time I did not realize that it was caused by the knx reconnection problem. I rather had the suspicion that my newly introduced Denon or Sony was guilty, or rather myself with completely mad rules and charts...which I have sorted out now.

**Comment by kaikreuzer** [comment]
If I can read the calendar correctly, March 2017 was a few months AFTER Oct 2016, so it would fit that this change introduced the issue.

**Comment by ThKern** [comment]
Stupid me. March 2017 is later than Oct 2016. Forgive me, I was only looking at the Month...

**Comment by ThKern** [comment]
...and this explains why I never had the problem under 1.x, as I kept on using org.openhab.binding.knx-1.7.0 from May 2015. So Kai is right.

**Comment by ThKern** [comment]
For those interested: for me the restart-workaround seems to work. I did it like that:
ssh into device
cd /etc/cron.daily
sudo nano restart_openhab
--> enter 
```

#!/bin/sh
#
# cron script to restart openhab.
#

systemctl restart openhab2.service

```
save and quit
sudo chmod a+x restart_openhab.
Then maybe sudo nano /etc/crontab and change the execution time of the daily jobs from 6:25 to 3:25, as at 6:25 does not seem a good time anyway for clean up purposes, as some people are awake at this time already. 
I hope this change survives future updates.

**Comment by lewie** [comment]
@kaikreuzer, I have searched (1x openHAB1 and 1x openHAB2) all my logs back to jun 2016. But no connection loose with OutOfMemoryError or similar in context with knx. 

The reason is probably that I use only ROUTER connections. Routing has no defined connection that could be lost. 

If it can be called a workaround? I recommend operating the KNX binding in ROUTER mode. ROUTER Hardware or openHAB> ROOTING> KNXD which tunnels to Hardware Interface. (MDT SCN-IP100.02 is a router and interface || MDT SCN-IP000.02 is a Interface only for tunneling)

@sjka, for the new KNX2 binding is still used [calimero 2.3](https://github.com/openhab/openhab2-addons/tree/9bf693e4dab65fff64b308e1908aac771897eacc/addons/binding/org.openhab.binding.knx/lib). 
Also facing this problem, with the conversion from calimero to java 8 ([since 2.4](https://calimero-project.github.io/)), significant improvements have been made and some known bugs have been fixed. In addition, the API has also changed significantly in some places. I'm not sure, but I think, the objection that there is no beta/release for 2.4 can not be considered as high in the case of calimero, as calimero evolves more evolutionarily. [I asked this Boris Malinowsky](https://github.com/calimero-project/calimero-core/issues/56)

**Comment by lewie** [comment]
@9037568, Chris why should this bugreport be closed? It is not fixed.
@ThKern, this is a really hard core worst case workaround... ;-)

**Comment by ThKern** [comment]
Yep. hard core.  :(
I am afraid my IP interface 730 by [Weinzierl ](https://www.weinzierl.de/index.php/de/alles-knx1/knx-devices/produktarchiv/knx-ip-interface-730) only supports tunnel. 
@lewie, you mean I should introduce [knx-daemon ](https://github.com/knxd/knxd) in between? Do you have a guide how to establish that? I thought I do not need this software, as I have a hardware interface?
Thanks
Thomas

**Comment by lewie** [comment]
@ThKern, 

Normally it is a one liner to connect knxd to your Interface. The other end of KNXD acts then as ROUTER which communicates with your openHAB.


**KNX IP Router** is for [ROUTER or/and TUNNEL Mode ](https://www.weinzierl.de/index.php/de/alles-knx1/knx-devices/knx-ip-router-751-de)
Configure KNX interfaces as ROUTER means, that they all communicate over a multicast IP address (default 224.0.23.12) together. This means, every ROUTER sends and listens at this IP without a hardwired connection.



**KNX IP Interface** is for [TUNNEL Mode only](https://www.weinzierl.de/index.php/de/alles-knx1/knx-devices/produktarchiv/knx-ip-interface-730)
In opposite, TUNNEL connections are hardwired to one endpoint -> A KNX interfaces as TUNNEL has a hardwired connection to a specific other KNX interface and its specific IP.






Here you find several information for a KNXD between layer:
https://community.openhab.org/t/oh2-not-working-with-knxd/16522/12
https://github.com/knxd/knxd/blob/master/systemd/knxd.conf

It would not be perfect but a workaround.

**Comment by 9037568** [comment]
@lewie: From what I've read so far, there's no bug here to be closed. Apparently there are some platforms out there that have connectivity issues that lead to memory problems.

But I have no dog in the fight. I'm ok with letting this sit here.

**Comment by kaikreuzer** [comment]
I also think we should leave the issue open as we require an upgrade of the Calimero lib to fix this issue. So let's see what kind of answer @lewie gets on https://github.com/calimero-project/calimero-core/issues/56.

**Comment by muenning** [comment]
@lewie I am using the MDT SCN-IP100.02 as written above but ROUTER mode does not work. As this seemed to me to be a common problem (as written in knx.cnf:
```
# Note: If you cannot get the ROUTER mode working (even if it claims it is connected), 
# use TUNNEL mode instead with setting both the ip of the KNX gateway and the localIp.
```
)
I spent some time trying to get it working but as I didn't find anything I can do I am stuck with TUNNEL mode for the time being. If you have any hint what to do to get ROUTER mode working, any advice is appreciated.

@kaikreuzer I'm not sure this is only a calimero issue. Maybe the initial connection loss is from a calimero bug but the following excessive re-connection attempts and infinitely spawning threads seems to be more a KNX binding issue. As for some reason your connection is stable the bug may be never triggered in your setup but mine has a lot of traffic which, besides the MDT hardware, may be hitting the bug. The traffic theory is supported by the fact that OH 2.x gets stuck about twice as fast as OH 1.x as it still resends the KNX events back to the bus despite the `ignorelocalevents=true` setting. This can be monitored easily with ETS.

I did some code digging and found something that might be the thread spawning loop. Java is one of my less used languages so please excuse imprecision in my wording. In `KNXConnection.java` the `connect` function sets up a listener to CloseEvent which in case the link is closed and `sAutoReconnectPeriod > 0` sets up a timer `TimerTask timerTask = new ConnectTimerTask(timer);` which is in fact a new thread. The `ConnectTimerTask` is calling `connect()` but terminates only in case of connect success. But connect schedules this task before the connection attempt is made and in case the connect it is not successful (like when the link slots are exhausted) the first task remains running and second one is spawned. And each one generates new threads on its own which explains the geometric increase in threads one the bug is triggered. I don't know when calimero triggers the CloseEvent but it seems perfectly sound to me to do it also on an unsuccessful connect. What is your opinion?

After this finding I tried with `autoReconnectPeriod=0` in my knx.cnf and the thread count seems stable for now about 24h. This is the longest stable run ever and the KNX router connections are not exhausted. Disabling reconnect seems a bad idea to me, referring to the parameter comment, but for some reason there are no lost KNX events so the communication is still working despite the disconnects, which are now about 15-40 min. apart and not up to several per second as before. The log also shows that despite the disabled reconnect OH reconnects almost instantly:
```
2018-01-20 23:17:21.221 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 10.0.0.160:3671 TP1 medium, device 15.15.244, hopcount 6)
2018-01-20 23:17:21.223 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-20 23:17:21.226 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-20 23:17:53.656 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 10.0.0.160:3671 TP1 medium, device 15.15.244, hopcount 6)
2018-01-20 23:17:53.661 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-20 23:17:53.662 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-20 23:43:23.178 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 10.0.0.160:3671 TP1 medium, device 15.15.244, hopcount 6)
2018-01-20 23:43:23.183 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-20 23:49:36.798 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 10.0.0.160:3671 TP1 medium, device 15.15.244, hopcount 6)
2018-01-20 23:49:36.800 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-20 23:49:36.803 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
```
I don't know what triggers the reconnect as I'm not sending events to the bus at that time but maybe the (undesired) event re-transmission is doing that. In this case that would explain the faster thread spawning in OH 2.x but also shows the need of a flag to suppress the (re-)connection when sending events if a reconnect timer is already running. Or cancel a reconnect timer if reconnected otherwise. Even without the re-transmission in OH 2.x an event could be sent at any time and in case of an already running timer (and a disconnected line) it would trigger a concurrent connect sequence.

@ThKern If interested, I got a hybrid 1.x - 2.x setup so I can easily switch between versions even with preserving persistence. Only the services has to be configured separately and new 2.x stuff (like the sitemap Element "Default") can not be used but the first do not change that much. If interested I can give you more information. In the meantime you may try if `autoReconnectPeriod=0` works for you, too.

**Comment by ThKern** [comment]
In the moment I am not unhappy with my Holzhammer described above.
AFAIR I had the period setting = 0 when I had openHAB installed on my Synology, but this does not work for me, as the knx addresses are highly integrated and once the connection to knx is lost my whole openHAB would be useless.
My problem is with all your nice support and workaround sugestions: It takes 36h+ until the problems start in my installation (with the restart every 24h my log stays clean from the reconnect messages). So quickly trying out a new workaround takes some time.

**Comment by muenning** [comment]
As said above and shown by the log, with `autoReconnectPeriod=0` on OH 2.x the lost connection is reestablished within 6 milliseconds, I think this short loss would affect your installation less than a reconnect timeout of some seconds. If it works in your setup the same way it does on mine. So if you try this parameter and check your logs it would be interesting to see if you also have the instant reconnect.
As on my setup the trouble starts usually minutes after starting OH 2.x and the systems gets unusable within 12h (more within 8h) I would need to restart about every 4h and that frequent restarts would mean a lot of dead time. Not a good solution for me.
FTR: With the changed parameter the thread count is still constant and the system works stable, I have not restarted it since then.

**Comment by ThKern** [comment]
Ok, I'm giving it a try with autoReconnectPeriod=0. Let's see the outcome in a few days (?). Thanks anyway

**Comment by ThKern** [comment]
I can confirm that it works good with autoReconnectPeriod=0. My Log:

```
2018-01-23 06:13:19.090 [WARN ] [tuwien.auto.calimero                ] - KNXnet/IP Tunneling 192.168.178.19:3671: received service acknowledgment with wrong send sequence 240, expected 241 - ignored
2018-01-23 06:13:19.097 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: maximum send attempts on object tunneling link link (closed) 192.168.178.19:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-23 06:13:19.104 [ERROR] [tuwien.auto.calimero                ] - KNXnet/IP Tunneling 192.168.178.19:3671: close connection - maximum send attempts
tuwien.auto.calimero.exception.KNXAckTimeoutException: maximum send attempts, no service acknowledgment received
	at tuwien.auto.calimero.knxnetip.ConnectionBase.send(ConnectionBase.java:261) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.knxnetip.KNXnetIPTunnel.send(KNXnetIPTunnel.java:149) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.link.KNXNetworkLinkIP.onSend(KNXNetworkLinkIP.java:263) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.link.AbstractLink.send(AbstractLink.java:304) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.link.KNXNetworkLinkIP.sendRequestWait(KNXNetworkLinkIP.java:240) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.process.ProcessCommunicatorImpl.write(ProcessCommunicatorImpl.java:466) [252:org.openhab.binding.knx:1.11.0]
	at tuwien.auto.calimero.process.ProcessCommunicatorImpl.write(ProcessCommunicatorImpl.java:438) [252:org.openhab.binding.knx:1.11.0]
	at org.openhab.binding.knx.internal.bus.KNXBinding.writeToKNX(KNXBinding.java:149) [252:org.openhab.binding.knx:1.11.0]
	at org.openhab.binding.knx.internal.bus.KNXBinding.internalReceiveUpdate(KNXBinding.java:126) [252:org.openhab.binding.knx:1.11.0]
	at org.openhab.core.binding.AbstractBinding.receiveUpdate(AbstractBinding.java:113) [251:org.openhab.core.compat1x:2.2.0]
	at org.openhab.core.events.AbstractEventSubscriber.handleEvent(AbstractEventSubscriber.java:39) [251:org.openhab.core.compat1x:2.2.0]
	at org.apache.felix.eventadmin.impl.handler.EventHandlerProxy.sendEvent(EventHandlerProxy.java:415) [3:org.apache.karaf.services.eventadmin:4.1.3]
	at org.apache.felix.eventadmin.impl.tasks.HandlerTask.runWithoutBlacklistTiming(HandlerTask.java:82) [3:org.apache.karaf.services.eventadmin:4.1.3]
	at org.apache.felix.eventadmin.impl.tasks.SyncDeliverTasks.execute(SyncDeliverTasks.java:104) [3:org.apache.karaf.services.eventadmin:4.1.3]
	at org.apache.felix.eventadmin.impl.tasks.AsyncDeliverTasks$TaskExecuter.run(AsyncDeliverTasks.java:166) [3:org.apache.karaf.services.eventadmin:4.1.3]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
2018-01-23 06:13:19.111 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-23 06:13:19.118 [WARN ] [.binding.knx.internal.bus.KNXBinding] - Value '2018-01-23T06:13:16' could not be sent to the KNX bus using datapoint 'state DP 2/3/1 Uhrzeit, DPT main 0 id 10.001, low priority' - retrying one time: maximum send attempts, no service acknowledgment received
2018-01-23 06:13:19.768 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on 192.168.178.19:3671 in mode TUNNEL.
```
A few hours later similar:
```
2018-01-23 15:16:50.561 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.178.19:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-23 15:16:50.565 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-23 15:16:50.624 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on 192.168.178.19:3671 in mode TUNNEL.
```
So I must first say Thanks for the good hint, and ask myself why I set autoReconnectPeriod=30 some (long) time ago. Maybe someone with more insight can explain why there is this `Received detach Event`. There is nothing in the log surrounding it.

**Comment by muenning** [comment]
Thanks for trying, I'm sure you could disable your Holzhammer now. For me it still works without problems and without restarts at a (slightly varying but in average) constant thread count.
Probably you have set this parameter for the same reason I did - according to the comment in knx.cfg and the binding docs setting autoReconnectPeriod > 0 is a good idea:
```
Seconds between connect retries when KNX link has been lost
0 means never retry, it will only reconnect on next write or read request.
Note: without periodic retries all events will be lost up to the next read/write request
```
So you result confirms the need to rework the connect/disconnect/reconnect handling in KNX binding and either fix the connection tracking or the reconnect sequence. My suggestion would be instead of launching a timer on each connect attempt to have one single timer launched on autoReconnectPeriod > 0 which periodically checks the connection status and in case of a disconnect it is the one and only place to reestablish the connection.
Of course if the mechanics of the currently observed instant reconnect is understood and it is reliable and not only due to the resend events bug then this would be better than my suggestion above and the autoReconnectPeriod parameter could be omitted, but I am afraid it wouldn't be reliable once the other bugs are fixed.

**Comment by lewie** [comment]
https://github.com/calimero-project/calimero-core/issues/56#issuecomment-361375387

Hmm, for me, it sounds 2.4 master is stable enough, is the most usable branch for us? There and only there will be go important improvements, 2.3 is past. 

It seams, Simons Issue comment will be fixed only in 2.4: https://github.com/calimero-project/calimero-core/issues/57#issue-291929017

Am I right, or how do you see that @kaikreuzer, @sjka?

**Comment by sjka** [comment]
I'd always prefer using released versions instead of a snapshot version unless for a very good reason - not because snapshots are inherently unstable, but rather because that always tends to create confusion, especially as soon as two or three snapshots of the same version get involved it will get hard to keep track of which fix/change made it into which version. 

The threading-issue is not necessarily a reason to force us to update. It can be worked around easy enough. Therefore I'd suggest to wait a for the release.

**Comment by lewie** [comment]
@sjka, it's your decision!

**Comment by 9037568** [comment]
Obsoleted by OH2 addon.
